utils/obsproc/NetCDFToIodaConverter.h

```diff
@@ -58,57 +65,108 @@ namespace gdasapp {
+      oops::Log::debug() << "--- all obsVal: " << iodaVarsAll.obsVal << std::endl;
+      oops::Log::debug() << "--- all obsError: " << iodaVarsAll.obsError << std::endl;
+      oops::Log::debug() << "--- all preQc: " << iodaVarsAll.preQc << std::endl;
+
+      // Create empty group backed by HDF file
+      if (oops::mpi::world().rank() == 0) {
+        ioda::Group group =
+          ioda::Engines::HH::createFile(
+                                        outputFilename_,
+                                        ioda::Engines::BackendCreateModes::Truncate_If_Exists);
+
+        // Update the group with the location dimension
+        ioda::NewDimensionScales_t
+          newDims {ioda::NewDimensionScale<int>("Location", iodaVars.location)};
+        ioda::ObsGroup ogrp = ioda::ObsGroup::generate(group, newDims);
+
+        // Set up the creation parameters
+        ioda::VariableCreationParameters float_params = createVariableParams<float>();
+        ioda::VariableCreationParameters int_params = createVariableParams<int>();
+
+        // Create the IODA variables
+        ioda::Variable adtIodaDatetime =
+          ogrp.vars.createWithScales<float>("Metadata/dateTime",
+                                            {ogrp.vars["Location"]}, float_params);
+        // TODO(All): Decide on what to use for the Epoch date
+        adtIodaDatetime.atts.add<std::string>("units", {"seconds since 9999-04-15T12:00:00Z"}, {1});
+
+        ioda::Variable adtIodaObsVal =
+          ogrp.vars.createWithScales<float>("ObsValue/"+variable_,
+                                            {ogrp.vars["Location"]}, float_params);
+        ioda::Variable adtIodaObsErr =
+          ogrp.vars.createWithScales<float>("ObsError/"+variable_,
+                                            {ogrp.vars["Location"]}, float_params);
+
+        ioda::Variable adtIodaPreQc =
+          ogrp.vars.createWithScales<int>("PreQC/"+variable_,
+                                            {ogrp.vars["Location"]}, int_params);
+
+        // Write adt obs info to group
+        adtIodaObsVal.writeWithEigenRegular(iodaVars.obsVal);
+        adtIodaObsErr.writeWithEigenRegular(iodaVars.obsError);
+        adtIodaPreQc.writeWithEigenRegular(iodaVars.preQc);
+      }
     }
 
    private:
     // Virtual method that reads the provider's netcdf file and store the relevant
     // info in a IodaVars struct
     virtual void providerToIodaVars(const std::string fileName, gdasapp::IodaVars & iodaVars) = 0;
 
+    // Gather for eigen array
+    template <typename T>
+    void gatherObs(const eckit::mpi::Comm & comm,
+                   const Eigen::Array<T, Eigen::Dynamic, 1> & obsPe,
+                   Eigen::Array<T, Eigen::Dynamic, 1> & obsAllPes) {
+      std::vector<T> tmpVec(obsPe.data(), obsPe.data() + obsPe.size());
+      oops::mpi::allGatherv(comm, tmpVec);
+      for (int i = 0; i < tmpVec.size(); ++i) {
+        obsAllPes(i) = tmpVec[i];
+      }
+    }
```

Comment:
Is this method needed? See previous comments...

Reply:
oops::mpi::allGatherv isn't overloaded for input Eigen::Array. This is my short cut to not repeat that chunk of code, short of adding an option in oops::mpi::allGatherv. Is there a better way to do this @rmclaren ?

Reply:
My MPI is a little rusty (its been a little while), so please forgive me if I say something stupid (I'm not used to thinking this way). It's coming back to me slowly so please bare with me :)

Having said that I don't think this code will work the way you think it does. For one thing the tmpVec is too small if there are more than one MPI tasks reading files (should be size of sum of nobs you computed in `reduce` i think)... I think there is also a lot of unecessary copying, and since the offset always starts at 0 (`int i = 0; `) you are overwriting data (only ending up with the data associated with the very last MPI tasks file). I don't think this method is really necessary at all...

Here is my suggestion (but please keep in mind that my MPI is rusty...).

```
      // Get the total number of obs across pe's
      comm.allReduce(nobs, nobs, eckit::mpi::sum());
      oops::Log::debug() << " my rank : " << myrank
                         << " Num pe's: " << comm.size()
                         << " nobs: " << nobs << std::endl;
      gdasapp::IodaVars iodaVarsAll(nobs);

      // Create empty group backed by HDF file
      if (oops::mpi::world().rank() == 0) {

        // Gather obsVal's
        oops::mpi::allGatherv(comm, iodaVarsAll.obsVal);
        oops::mpi::allGatherv(comm, iodaVarsAll.obsError);
        oops::mpi::allGatherv(comm, iodaVarsAll.obs.preQc);
  
        oops::Log::debug() << "--- all nobs: " << iodaVarsAll.obsVal.size() << std::endl;
        oops::Log::debug() << "--- all obsVal: " << iodaVarsAll.obsVal << std::endl;
        oops::Log::debug() << "--- all obsError: " << iodaVarsAll.obsError << std::endl;
        oops::Log::debug() << "--- all preQc: " << iodaVarsAll.preQc << std::endl;

        ioda::Group group =
          ioda::Engines::HH::createFile(
                                        outputFilename_,
                                        ioda::Engines::BackendCreateModes::Truncate_If_Exists);

// ***** unchanged beyond this point ******//
```

Reply:
hold on a second

Reply:
I think what I did is correct @rmclaren , but maybe not amazingly efficient. It does properly concatenate the input files.
The oops mpi call is here:
https://github.com/JCSDA-internal/oops/blob/c03558814bb8cb81de71265b98f7e54ef0d953f9/src/oops/mpi/mpi.h#L219

The size of x is changed in that function.

What I might change in the future is to all gather "in place" instead of creating a new ```iodaVarsAll```.

Reply:
the size of ```tmpVec``` changes in line 152 @rmclaren .

Reply:
Ok, I looked at the oops implementation for allGatherv in oops and its actually swapping out the data buffer under the covers via a std::move...  I think what you had originally works as written, but could be improved.

I think you might not need `iodaVarsAll` at all. This means you don't need to do the `allReduce`. So you could just do the gather and then write the iodaVar values.


```cxx

       // Gather obsVal's
      oops::mpi::allGatherv(comm, iodaVar.obsVal);
      oops::mpi::allGatherv(comm, iodaVars.obsError);
      oops::mpi::allGatherv(comm, iodaVarsAll.preQc);

      oops::Log::debug() << "--- all nobs: " << iodaVars.obsVal.size() << std::endl;
      oops::Log::debug() << "--- all obsVal: " << iodaVars.obsVal << std::endl;
      oops::Log::debug() << "--- all obsError: " << iodaVars.obsError << std::endl;
      oops::Log::debug() << "--- all preQc: " << iodaVars.preQc << std::endl;
   
      // Create empty group backed by HDF file
      if (oops::mpi::world().rank() == 0)
      {
        ioda::Group group =
          ioda::Engines::HH::createFile(
                                        outputFilename_,
                                        ioda::Engines::BackendCreateModes::Truncate_If_Exists);

// ***** end of suggestion ******//
```

Reply:
Just a side note, instead of manually copying the data in `gatherObs` (in the for loop) you could have just done a std::move and "steal" the pointer.

Sorry for my confusions...